Add Log Replication Cluster Role ITs

zhangn49 opened the pull request:
## Overview

Description:

Add Log Replication Cluster Role ITs

* Modified Default Cluster Manager to support process-based ITs
* Removed current CorfuReplicationSiteConfigIT, since it won't work on Travis.

Why should this be merged: 

Related issue(s) (if applicable): #2627 


## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

zhangn49 commented on infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationDiscoveryService.java:
```diff
@@ -578,15 +581,18 @@ private void onStandbyClusterAddRemove(TopologyDescriptor discoveredTopology) {
         }
 
         updateLocalTopology(discoveredTopology);
+        updateReplicationManagerTopology(discoveredTopology);
+        // Update topology config id in metadata manager
+        logReplicationMetadataManager.setupTopologyConfigId(topologyDescriptor.getTopologyConfigId());
+        log.debug("Persist new topologyConfigId {}, cluster id={}, status={}", topologyDescriptor.getTopologyConfigId(),
+                localClusterDescriptor.getClusterId(), localClusterDescriptor.getRole());
```
Even if current node's role did not change, it does not guarantee that others' role or config id is not changed...  
So updating sender/receiver's config id is necessary here.

zhangn49 commented on infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationManager.java:
```diff
@@ -177,35 +179,41 @@ private void stopLogReplicationRuntime(String remoteClusterId) {
     public void processStandbyChange(TopologyDescriptor newConfig) {
         if (newConfig.getTopologyConfigId() != topology.getTopologyConfigId()) {
-            log.error("Detected changes in the topology. The new topology descriptor {} doesn't have the same " +
+            log.warn("Detected changes in the topology. The new topology descriptor {} doesn't have the same " +
```
We do not assume that config id change implies a role change, so it is possible that we have a higher config id.

xiaoqin2012 replied:
We may add more comments to the function to explain that this function only is executed if the configID doesn't change.

zhangn49 replied:
Ideally, there is no config id change when we only processStandbyChange. I will add a short comment.

zhangn49 commented on test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java:
```diff
@@ -93,104 +91,520 @@ public void testLogReplicationRoleFlip() throws Exception {
+     * 5. Standby will drop messages and keep size 0
+     * 6. Verify active map becomes size 0, since source size is 0
+     */
+    //@Test
```
Since there is no way to make sure apply completes before the writer reset, we do not support DuringTransferPhase & DuringApplyPhase now.

xiaoqin2012 commented:
Just curious, all tests are passing or there are still issues?

zhangn49 commented:
> Just curious, all tests are passing or there are still issues?

All 4 tests passed. 
1.testNewConfigWithSwitchRole
2.testNewConfigWithTwoActive
3.testNewConfigWithAllStandby
4.testNewConfigWithInvalidClusters

xiaoqin2012 commented on infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationDiscoveryService.java:
```diff
@@ -578,15 +581,18 @@ private void onStandbyClusterAddRemove(TopologyDescriptor discoveredTopology) {
+        if (localClusterDescriptor.getRole() == ClusterRole.ACTIVE) {
+            if (replicationManager != null && isLeader.get()) {
+                replicationManager.processStandbyChange(discoveredTopology);
+            }
```
If the topologyConfigId has been change, it will overrule the processStandbyChange. processStandbyChange only is called when the topologyConfigId doesn't change and the list of standbys have been changed. 
Is the logic correct that processStandbyChange called before updating the localTopology?

zhangn49 replied:
Since we don't assume that config id change implies a role change anymore, processStandbyChange won't be overruled. Now we only detect if local node role changes or not, then perform `onClusterRoleChange`  or `onStandbyClusterAddRemove`



> Is the logic correct that processStandbyChange called before updating the localTopology?

I think the logic is correct, since stopping old replication relies on old config.

xiaoqin2012 commented on test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java:
```diff
@@ -93,104 +91,520 @@ public void testLogReplicationRoleFlip() throws Exception {
+
+        // Wait until data is fully replicated
+        waitForReplication(size -> size == firstBatch, mapStandby, firstBatch);
+        log.info("After full sync, both maps have size {}. Current active corfu[{}] log tail " +
```
Can we also add a check of the log replication metadata at the standby side. It should indicate a status of snapshot full sync has finished and the timestamp for the snapshot full sync should agree with the current active log tail.

zhangn49 replied:
Yeah, it makes sense to check metadata. I need to add helper functions to do some Corfu table queries...
I will make sure this PR's logic is correct, then I will have a look at metadata tables and add extra checks.

annym commented on infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultClusterManager.java:
```diff
@@ -41,28 +57,72 @@
+    public static final String CONFIG_NAMESPACE = "ns_lr_config_it";
+    public static final String CONFIG_TABLE_NAME = "lr_config_it";
+    public static final CommonTypes.Uuid OP_RESUME = CommonTypes.Uuid.newBuilder().setLsb(0L).setMsb(0L).build();
+    public static final CommonTypes.Uuid OP_SWITCH = CommonTypes.Uuid.newBuilder().setLsb(1L).setMsb(1L).build();
```
What are these OP? can we have a general description?

zhangn49 replied:
It is operations. I think it is kind of ugly to hardcode it and I will try to directly write new config to tables.

annym commented on test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java:
```diff
@@ -0,0 +1,619 @@
+     * 9. Resume to standby and verify data is fully replicated again.
+     */
+    @Test
+    public void testNewConfigWithInvalidClusters() throws Exception {
```
If you take a look at PR #2640 I broke all these functions of writing to active/standby maps/ stopping into generic utilities available in LogReplicationAbstractIT, which would reduce a lot of the code here (just a suggestion)

pankti-m replied:
A similar suggestion is to move the common code from all tests to a common place in this file so the tests themselves are smaller

annym commented on test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java:
```diff
@@ -0,0 +1,619 @@
+
+    /**
+     * This test verify config change with a role switch.
+     * <p>
```
These tests are good! I think one that we might be missing and could be interesting is:
- Snapshot and Log Entry (check it completes)
- Stop both LRs
- Enforce Role Switch
- Bring them back

This way we can verify it performs correctly the site flip even when coming back up after with a new config.

zhangn49 commented on infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationDiscoveryService.java:
```diff
@@ -510,6 +510,9 @@ public void onClusterRoleChange(TopologyDescriptor newTopology) {
                 localClusterDescriptor.getClusterId(), localClusterDescriptor.getRole());
 
+        // Update replication manager
+        updateReplicationManagerTopology(newTopology);
```
Let's assume that we start with topology[0]: {active: A, standby:B}, then update with topology[1]:{standby: A, B}. If we do not update replication manager topology, cluster B will have outdated epoch 0.

Next time a new active C comes in, topology is [2]{active C, standby:A, B}. For cluster B, its role never changed, and it won't update replication manager topology.

zhangn49 commented on infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationDiscoveryService.java:
```diff
@@ -578,15 +581,18 @@ private void onStandbyClusterAddRemove(TopologyDescriptor discoveredTopology) {
         }
 
         updateLocalTopology(discoveredTopology);
+        updateReplicationManagerTopology(discoveredTopology);
```
Just like my example above, we need to keep replication manager's topology epoch updated.

zhangn49 replied:
@pankti-m  Yes, please see my example at line 515

zhangn49 commented on infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultClusterManager.java:
```diff
@@ -41,28 +57,72 @@
+    public static final CommonTypes.Uuid OP_INVALID = CommonTypes.Uuid.newBuilder().setLsb(4L).setMsb(4L).build();
+
+    @Getter
+    private long epoch;
```
It is configId. Renamed.

zhangn49 commented on infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultClusterManager.java:
```diff
@@ -41,28 +57,72 @@
+            );
+            table.clear();
+        } catch (Exception e) {
+            // Ignore
```
Yes, it should be thrown and discovery service will timeout and fail the test.

zhangn49 commented on test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java:
```diff
@@ -0,0 +1,619 @@
+
+        // Double check after 10 seconds
+        TimeUnit.SECONDS.sleep(mediumInterval);
+        //assertThat(mapActive.size()).isEqualTo(thirdBatch);
```
Yes, it works now. Forgot to bring it back after debugging.

zhangn49 commented on test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java:
```diff
@@ -0,0 +1,619 @@
+                standbyRuntime.getAddressSpaceView().getLogTail());
+
+        // Verify data
+        for (int i = 0; i < secondBatch; i++) {
```
Yeah, it is cleaner. I think for loop can make sure all entries exist and help understand how many entries should be there.

zhangn49 commented on infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultClusterManager.java:
```diff
@@ -181,57 +229,151 @@ public TopologyConfigurationMsg queryTopologyConfig(boolean useCached) {
+        currentConfig.getActiveClusters().values().forEach(activeCluster ->
+                newStandbyClusters.add(new ClusterDescriptor(activeCluster, ClusterRole.STANDBY)));
+        currentConfig.getStandbyClusters().values().forEach(standbyCluster ->
+                newActiveClusters.add(new ClusterDescriptor(standbyCluster, ClusterRole.ACTIVE)));
```
Yes, done.

zhangn49 commented on infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultClusterManager.java:
```diff
@@ -181,57 +229,151 @@ public TopologyConfigurationMsg queryTopologyConfig(boolean useCached) {
         }
     }
 
+    public static class ConfigStreamListener implements StreamListener {
```
Done.

zhangn49 commented on infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultClusterManager.java:
```diff
@@ -181,57 +229,151 @@ public TopologyConfigurationMsg queryTopologyConfig(boolean useCached) {
         }
     }
 
+    public static class ConfigStreamListener implements StreamListener {
```
Yes, it is not required.

pankti-m replied:
why do we need to use the table and a listener on it?  The test can just iterate thru all cases and inject different topologies right?

zhangn49 replied:
@pankti-m Listener enables us not to busy-wait.

pankti-m replied:
... not sure if I understand, but we can discuss this offline.  It is not a functional issue.

zhangn49 commented:
@annym @xiaoqin2012 
Hi, I have addressed most of your comments. And I think it is good to have this PR merged before the cut, so I can test role change cases in the production environment.

I will have a separate PR to address 3 things later:
1. Code cleanup: Remove ClusterManagerCallback, change config operations, and reuse functions in #2640 
2. More test cases: 'Stop LR, Role switch, and Bring back', as well as two cases I commented out.
3. Add helper functions to query and check standby metadata.

zhangn49 commented:
Pending items will be tracked by #2656

pankti-m commented on infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultClusterManager.java:
```diff
@@ -112,9 +171,8 @@ public static TopologyDescriptor readConfig() throws IOException {
-        } catch (FileNotFoundException e) {
-            log.warn("Plugin Config File {} does not exist. Using default configs", config_file);
+        } catch (IOException e) {
+            log.warn("Plugin Config File {} does not exist. Using default configs", CONFIG_FILE_PATH);
```
this can be an inaccurate error message if the exception was not FileNotFound.  Why was the exception type changed?

zhangn49 replied:
Auto close will throw IOException. I think it is fine, since the entire default cluster manager is test only.

pankti-m commented on infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultClusterManager.java:
```diff
@@ -181,57 +229,161 @@ public TopologyConfigurationMsg queryTopologyConfig(boolean useCached) {
     /**
-     * Enforce one of the standby Cluster's to become the new active cluster and current active to become standby
+     * Create a new topology config, which changes one of the standby as the active,
+     * and active as standby. Data should flow in the reverse direction.
```
can you annotate it with @TestOnly or add it in the comment that this is only used by tests?

zhangn49 replied:
Actually the entire DefaultClusterManager is test-only.

pankti-m commented on infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultClusterManager.java:
```diff
@@ -181,57 +229,161 @@ public TopologyConfigurationMsg queryTopologyConfig(boolean useCached) {
+     * Create a new topology config, which marks all standby cluster as active on purpose.
+     * System should drop messages between any two active clusters.
+     **/
+    public TopologyDescriptor generateConfigWithAllActive() {
```
can you add an annotation or comment that this in only for testing?

zhangn49 replied:
The same. I think I will refactor this part in the next PR.

pankti-m commented on infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultClusterManager.java:
```diff
@@ -181,57 +229,161 @@ public TopologyConfigurationMsg queryTopologyConfig(boolean useCached) {
+     * Create a new topology config, which marks all standby cluster as invalid on purpose.
+     * System should not send messages in this case.
+     **/
+    public TopologyDescriptor generateConfigWithInvalid() {
```
maybe we should also add a case with active also as invalid

zhangn49 replied:
Sounds good, will add it in the next PR.

pankti-m commented on infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/StreamsSnapshotWriter.java:
```diff
@@ -4,6 +4,7 @@
 import lombok.Getter;
 import lombok.extern.slf4j.Slf4j;
 import org.corfudb.infrastructure.logreplication.LogReplicationConfig;
+import org.corfudb.infrastructure.logreplication.proto.LogReplicationMetadata.LogReplicationMetadataVal;
```
is this import needed?

zhangn49 replied:
Done.

pankti-m commented on test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java:
```diff
@@ -0,0 +1,618 @@
+ * Each test will start with two single node corfu servers, and two single node log replicators.
+ */
+@Slf4j
+@SuppressWarnings("checkstyle:magicnumber")
```
do you plan to remove this?

zhangn49 replied:
I will keep this one, since it is an IT class.

pankti-m commented on test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java:
```diff
@@ -0,0 +1,618 @@
+     * 2. Write 50 entries to active map
+     * 3. Start log replication: Node 9010 - active, Node 9020 - standby
+     * 4. Perform a role switch with corfu store
+     * 5. Standby will drop messages and keep size 0
```
Actually, role is not expected to change when in transfer phase.  Only when we report transfer as 100%, the flip happens.  Will have to decide how we want to handle this case.

zhangn49 replied:
Yeah, these two cases are not very clear right now. We did know much about workflow when designed these two cases.

pankti-m commented on test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java:
```diff
@@ -0,0 +1,618 @@
+     * 4. Wait for Snapshot Sync, both maps have size 10
+     * 5. Write 5 more entries to active map, to verify Log Entry Sync
+     * 6. Perform a two-standby config update with corfu store
+     * 7. Write 5 more entries to active map
```
actually, writes do not happen on the standby.  Technically they can, but we should decide if from Corfu we want to allow them

zhangn49 replied:
Good to know it

pankti-m commented on test/src/test/resources/logback-test.xml:
```diff
@@ -44,8 +44,8 @@
     </logger>
 
     <!--<root level="DEBUG">-->
-        <!--&lt;!&ndash;<appender-ref ref="FILE" />&ndash;&gt;-->
```
do you plan to revert this before merging?

zhangn49 replied:
I think the old one is double '<!--'. The new one is cleaner.

zhangn49 commented on infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationDiscoveryService.java:
```diff
@@ -489,11 +489,12 @@ public void processLockRelease() {
      *   - Potential cluster role change
      *
+     * Cluster change from active to standby is a two step process, we first confirm that
+     * we are ready to do the cluster role change, so by the time we receive cluster change
```
I guess we confirm it by query replication status.

zhangn49 commented on infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultClusterManager.java:
```diff
@@ -1,33 +1,49 @@
+ * one active cluster, and one or more standby clusters.
+ */
 @Slf4j
+@SuppressWarnings("checkstyle:magicnumber")
```
Removed.
